File: src/ios/trees/QtLocation.swift
import Foundation
import CoreLocation

struct QtLocation
{
    var latitude: QtCoordinate
    var longitude: QtCoordinate
}

extension QtLocation
{
    init(json: JsonObject) 
    {
        let latitude = json["latitude"] as! QtCoordinate
        let longitude = json["longitude"] as! QtCoordinate
        self.init(latitude: latitude, longitude: longitude)
    }
}

////////////////////////////////////////
// CLLocationCoordinate2D conversions

extension QtLocation
{
    init(_ location: CLLocationCoordinate2D) 
    {
        self.init(latitude: location.latitude, longitude: location.longitude)
    }
}

extension CLLocationCoordinate2D
{
    init(_ location: QtLocation)
    {
        self.init(latitude: location.latitude, longitude: location.longitude)
    }
}

File: src/android/app/src/main/java/com/crickettechnology/trees/QtLocation.java
package com.crickettechnology.trees;

import com.google.android.gms.maps.model.*;
import android.location.*;
import org.json.*;

class QtLocation
{
    public QtLocation(double latitude, double longitude)
    {
        this.latitude = latitude;
        this.longitude = longitude;
    }

    public QtLocation(Location location)
    {
        this.latitude = location.getLatitude();
        this.longitude = location.getLongitude();
    }

    public QtLocation(JSONObject json)
    {
        this.latitude = json.optDouble("latitude");
        this.longitude = json.optDouble("longitude");
    }

    public double latitude;
    public double longitude;

    public double distanceSquared(QtLocation location)
    {
        double dlat = this.latitude - location.latitude;
        double dlon = this.longitude - location.longitude;
        return dlat*dlat + dlon*dlon;
    }

    public double distance(QtLocation location)
    {
        return Math.sqrt(distanceSquared(location));
    }

    public LatLng toLatLng()
    {
        return new LatLng(this.latitude, this.longitude);
    }
}

